examples_tests/61.UI/UI_System.cpp

```diff
@@ -0,0 +1,1102 @@
+				0,
+				sizeof(ImDrawIdx) == 2 ? EIT_16BIT : EIT_32BIT
+			);
+
+			static constexpr size_t offset = 0;
+			commandBuffer.bindVertexBuffers(
+				0,
+				1,
+				&vertexBuffer.get(),
+				&offset
+			);
+
+			// Setup viewport:
+			SViewport const viewport
+			{
+				.x = 0,
+				.y = 0,
+				.width = frameBufferWidth,
+				.height = frameBufferHeight,
+				.minDepth = 0.0f,
+				.maxDepth = 1.0f,
+			};
+			commandBuffer.setViewport(0, 1, &viewport);
+
+			// Setup scale and translation:
+			// Our visible imgui space lies from draw_data->DisplayPps (top left) to draw_data->DisplayPos+data_data->DisplaySize (bottom right). DisplayPos is (0,0) for single viewport apps.
+			{
+				PushConstants constants{};
+				constants.scale[0] = 2.0f / drawData->DisplaySize.x;
+				constants.scale[1] = 2.0f / drawData->DisplaySize.y;
+				constants.translate[0] = -1.0f - drawData->DisplayPos.x * constants.scale[0];
+				constants.translate[1] = -1.0f - drawData->DisplayPos.y * constants.scale[1];
+
+				commandBuffer.pushConstants(
+					state->independentPipeline->getLayout(),
+					IShader::ESS_VERTEX,
+					0,
+					sizeof(constants),
+					&constants
+				);
+			}
+
+			// Will project scissor/clipping rectangles into frame-buffer space
+			ImVec2 const clip_off = drawData->DisplayPos;         // (0,0) unless using multi-viewports
+			ImVec2 const clip_scale = drawData->FramebufferScale; // (1,1) unless using retina display which are often (2,2)
+
+			// Render command lists
+			// (Because we merged all buffers into a single one, we maintain our own offset into them)
+			int global_vtx_offset = 0;
+			int global_idx_offset = 0;
+			for (int n = 0; n < drawData->CmdListsCount; n++)
+			{
+				const ImDrawList* cmd_list = drawData->CmdLists[n];
+				for (int cmd_i = 0; cmd_i < cmd_list->CmdBuffer.Size; cmd_i++)
+				{
+					const ImDrawCmd* pcmd = &cmd_list->CmdBuffer[cmd_i];
+
+					// Project scissor/clipping rectangles into frame-buffer space
+					ImVec4 clip_rect;
+					clip_rect.x = (pcmd->ClipRect.x - clip_off.x) * clip_scale.x;
+					clip_rect.y = (pcmd->ClipRect.y - clip_off.y) * clip_scale.y;
+					clip_rect.z = (pcmd->ClipRect.z - clip_off.x) * clip_scale.x;
+					clip_rect.w = (pcmd->ClipRect.w - clip_off.y) * clip_scale.y;
+
+					if (clip_rect.x < frameBufferWidth && clip_rect.y < frameBufferHeight && clip_rect.z >= 0.0f && clip_rect.w >= 0.0f)
+					{
+						// Negative offsets are illegal for vkCmdSetScissor
+						if (clip_rect.x < 0.0f)
+							clip_rect.x = 0.0f;
+						if (clip_rect.y < 0.0f)
+							clip_rect.y = 0.0f;
+
+						{// Apply scissor/clipping rectangle
+							VkRect2D scissor{};
+							scissor.offset.x = static_cast<int32_t>(clip_rect.x);
+							scissor.offset.y = static_cast<int32_t>(clip_rect.y);
+							scissor.extent.width = static_cast<uint32_t>(clip_rect.z - clip_rect.x);
+							scissor.extent.height = static_cast<uint32_t>(clip_rect.w - clip_rect.y);
+							commandBuffer.setScissor(0, 1, &scissor);
+						}
```

Comment:
an idea for optimization, use 4 programmable clip planes instead `gl_ClipDistance[0-4]` in the vertex shader of changing the scissor rect all the time.

Add a second vertex attribute with a per-instance divisor, which would be a vec4 containing the scissor rect in NDC.
(basically a linear buffer of `vec4` scissor rects scaled to be in [-1,1]^2 relative to th viewport)

Then have a third GPU buffer full of DrawElementsIndirect, always drawing one instance, but using `baseInstance` to identify which scissor rect you want to use.

Then draw everything with a single drawcall !